include/ada/url_pattern_helpers.h

```diff
@@ -0,0 +1,344 @@
+/**
+ * @file url_pattern_helpers.h
+ * @brief Declaration for the URLPattern helpers.
+ */
+#ifndef ADA_URL_PATTERN_HELPERS_H
+#define ADA_URL_PATTERN_HELPERS_H
+
+#include "ada/expected.h"
+#include "ada/url_pattern.h"
+
+#include <string>
+#include <tuple>
+#include <vector>
+
+namespace ada::url_pattern_helpers {
+
+// @see https://urlpattern.spec.whatwg.org/#token
+enum class token_type : uint8_t {
+  INVALID_CHAR,    // 0
+  OPEN,            // 1
+  CLOSE,           // 2
+  REGEXP,          // 3
+  NAME,            // 4
+  CHAR,            // 5
+  ESCAPED_CHAR,    // 6
+  OTHER_MODIFIER,  // 7
+  ASTERISK,        // 8
+  END,             // 9
+};
+
+std::string to_string(token_type type);
+
+// @see https://urlpattern.spec.whatwg.org/#tokenize-policy
+enum class token_policy {
+  STRICT,
+  LENIENT,
+};
+
+// @see https://urlpattern.spec.whatwg.org/#tokens
+struct Token {
+  // A token has an associated type, a string, initially "invalid-char".
+  token_type type = token_type::INVALID_CHAR;
+
+  // A token has an associated index, a number, initially 0. It is the position
+  // of the first code point in the pattern string represented by the token.
+  size_t index = 0;
+
+  // A token has an associated value, a string, initially the empty string. It
+  // contains the code points from the pattern string represented by the token.
+  std::string value{};
+};
+
+// @see https://urlpattern.spec.whatwg.org/#pattern-parser
+template <url_pattern_encoding_callback F>
+class url_pattern_parser {
```

Comment:
Instance of the template class url_pattern_parser appear to own tokens. We want to make sure that the token pointers never survive beyond the lifetime of their originator.